gateway/src/apicast/policy/token_introspection/token_introspection.lua

```diff
@@ -0,0 +1,68 @@
+local policy = require('apicast.policy')
+local _M = policy.new('Token Introspection Policy')
+
+local cjson = require('cjson')
+local http_authorization = require 'resty.http_authorization'
+local http_ng = require 'resty.http_ng'
+local user_agent = require 'apicast.user_agent'
+local resty_env = require('resty.env')
+
+local new = _M.new
+
+function _M.new(config)
+  local self = new()
+  self.config = config or {}
+  --- authorization for the token introspection endpoint.
+  -- https://tools.ietf.org/html/rfc7662#section-2.2
+  self.credential = 'Basic ' .. ngx.encode_base64(table.concat({ self.config.client_id or '', self.config.client_secret or '' }, ':'))
+  self.introspection_url = config.introspection_url
+  self.http_client = http_ng.new{
+    backend = config.client,
+    options = {
+      header = { ['User-Agent'] = user_agent() },
+      ssl = { verify = resty_env.enabled('OPENSSL_VERIFY') }
+    }
+  }
+  return self
+end
+
+--- OAuth 2.0 Token Introspection defined in RFC7662.
+-- https://tools.ietf.org/html/rfc7662
+local function introspect_token(self, token)
+  local opts = {
+    headers = {
+      ['Authorization'] = self.credential
+    }
+  }
+  --- Parameters for the token introspection endpoint.
+  -- https://tools.ietf.org/html/rfc7662#section-2.1
+  local res, err = self.http_client.post(self.introspection_url , { token = token, token_type_hint = 'access_token'}, opts)
+  if res and err then
+    ngx.log(ngx.WARN, 'token introspection error: ', err, ' url: ', self.introspection_url)
+    return { active = false }
+  end
+
+  if res.status == 200 then
+    local token_info = cjson.decode(res.body)
+    return token_info
```

Comment:
This can return `nil` when `cjson.decode` returns `nil, err`. That can happen for example when the response body is not valid json (html usually).

Reply:
Just when fixing https://github.com/3scale/apicast/pull/626 I realised there is plenty of error cases.

If the API would return `null` then this would be `ngx.null` and return the same error as in https://github.com/3scale/apicast/pull/626.  

This function should always return a table even when the json response can be a string, number, null, ...

And it should normalize `active` because `if ngx.null` is truthy so `{ "active": null }` would evaluate to true.